src/engraving/rendering/dev/beamtremololayout.cpp

```diff
@@ -96,40 +96,68 @@ void BeamTremoloLayout::offsetBeamToRemoveCollisions(const BeamBase* item, const
     double endY = (isStartDictator ? pointer : dictator) * ldata->spatium / 4 + tolerance;
+    // min stem lengths according to how many beams there are (starting with 1)
+    static const int minStemLengths[] = { 11, 13, 15, 18, 21, 24, 27, 30 };
 
+    const ChordRest* firstChordRest = chordRests.front();
+    const ChordRest* lastChordRest = chordRests.back();
+
+    size_t curChordRest = 0;
     for (ChordRest* chordRest : chordRests) {
         if (!chordRest->isChord() || chordRest == ldata->elements.back() || chordRest == ldata->elements.front()) {
+            curChordRest++;
             continue;
         }
 
-        PointF anchor = chordBeamAnchor(ldata, chordRest, ChordBeamAnchorType::Middle) - item->pagePos();
+        int sameLineException = 0;
 
-        int slope = std::abs(dictator - pointer);
-        double reduction = 0.0;
-        if (!isFlat) {
-            if (slope <= 3) {
-                reduction = 0.25 * ldata->spatium;
-            } else if (slope <= 6) {
-                reduction = 0.5 * ldata->spatium;
-            } else { // slope > 6
-                reduction = 0.75 * ldata->spatium;
+        if (chordRest->isChord()
+            && ((curChordRest == 1 && firstChordRest->isChord()) || (curChordRest == chordRests.size() - 2 && lastChordRest->isChord()))) {
+            const Chord* innerChord = toChord(chordRest);
+            const Chord* outerChord = curChordRest == 1 && firstChordRest->isChord() ? toChord(firstChordRest) : toChord(lastChordRest);
+            const int innerLine = ldata->up ? innerChord->upNote()->line() : innerChord->downNote()->line();
+            const int outerLine = ldata->up ? outerChord->upNote()->line() : outerChord->downNote()->line();
+
+            if (innerLine == outerLine) {
+                sameLineException = 1;
             }
         }
 
-        if (endX != startX) {
-            // avoid division by zero for zero-length beams (can exist as a pre-layout state used
-            // for horizontal spacing computations)
-            double proportionAlongX = (anchor.x() - startX) / (endX - startX);
+        PointF anchor = chordBeamAnchor(ldata, chordRest, ChordBeamAnchorType::Middle) - item->pagePos();
 
+        const int minLen = minStemLengths[std::max(strokeCount(ldata, chordRest) - 1, 0)] - sameLineException;
+
+        if (endX != startX) {
             while (true) {
-                double desiredY = proportionAlongX * (endY - startY) + startY;
-                bool beamClearsAnchor = (ldata->up && muse::RealIsEqualOrLess(desiredY, anchor.y() + reduction))
-                                        || (!ldata->up && muse::RealIsEqualOrMore(desiredY, anchor.y() - reduction));
-                if (beamClearsAnchor) {
+                // avoid division by zero for zero-length beams (can exist as a pre-layout state used
+                // for horizontal spacing computations)
+                const double proportionAlongX = (anchor.x() - startX) / (endX - startX);
```

Comment:
If this variable doesn't change within the loop it should be moved out, so we save having to recalculate it each time.